drivers/watchdog/wdt_nrfx_wdt.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <nrfx_wdt.h>
+#include <nrf_clock.h>
+#include <clock_control.h>
+#include <watchdog.h>
+
+#define SYS_LOG_DOMAIN "WDT NRF"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_WDT_LEVEL
+#include <logging/sys_log.h>
+
+#define NRF_WDT_CHANNELS_NUMBER 8
+
+static struct device DEVICE_NAME_GET(wdt_nrf);
+
+/* There are eight reload registers (channels) in Nordic's WDT. */
+static nrfx_wdt_channel_id m_ch[NRF_WDT_CHANNELS_NUMBER];
+
+/* Each activated and not reloaded channel can generate watchdog interrupt.
+ * Array m_cb provides storing callbacks for each channel.
+ */
+static wdt_callback_t      m_cb[NRF_WDT_CHANNELS_NUMBER];
+
+/*
+ * The m_alloc_ch_num variable stores number of currently allocated
+ * and activated watchdog channels.
+ */
+static u32_t               m_alloc_ch_num;
+
+
+static void wdt_event_handler(void)
+{
+	u32_t i;
+
+	for (i = 0; i < m_alloc_ch_num; ++i) {
+		if (nrf_wdt_request_status(m_ch[i])) {
+			if (m_cb[i]) {
+				m_cb[i](DEVICE_GET(wdt_nrf), &m_ch[i]);
+			}
+		}
+	}
+}
+
+static void wdt_nrf_isr(struct device *dev)
+{
+	ARG_UNUSED(dev);
+	/* We need to implement our own interrupt handler until nrfx one will
+	 * be fixed.
+	 * Clearing events also clears the reload register and has to be done
+	 * AFTER event handler.
+	 */
+	if (nrf_wdt_int_enable_check(NRF_WDT_INT_TIMEOUT_MASK)) {
+		wdt_event_handler();
+		nrf_wdt_event_clear(NRF_WDT_EVENT_TIMEOUT);
+	}
+}
+
+static int wdt_nrf_setup(struct device *dev, u8_t options)
+{
+	nrfx_wdt_config_t config;
+	nrfx_err_t err_code;
+
+	/* Set default value according to default value in MCU. */
+	config.behaviour = NRF_WDT_BEHAVIOUR_RUN_SLEEP;
+
+	ARG_UNUSED(dev);
+
+	if (!nrf_clock_lf_is_running()) {
+		return -ENOTSUP;
+	}
+
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		config.behaviour &= ~NRF_WDT_BEHAVIOUR_RUN_SLEEP;
+	}
+
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		config.behaviour &= ~NRF_WDT_BEHAVIOUR_RUN_HALT;
+	}
+
+	config.reload_value = CONFIG_WDT_NRF_TIMEOUT;
+
+	err_code = nrfx_wdt_init(&config, wdt_event_handler);
+	if (err_code == NRFX_ERROR_INVALID_STATE) {
+		return -EBUSY;
+	}
+
+	IRQ_CONNECT(CONFIG_WDT_NRF_IRQ,
+		    CONFIG_WDT_NRF_IRQ_PRI, wdt_nrf_isr,
+		    DEVICE_GET(wdt_nrf), 0);
+	irq_enable(CONFIG_WDT_NRF_IRQ);
+
+	nrfx_wdt_enable();
+
+	return 0;
+}
+
+static int wdt_nrf_disable(struct device *dev)
+{
+	/* Started watchdog cannot be stopped on nRF devices. */
+	ARG_UNUSED(dev);
+	return 0;
+}
+
+static int wdt_nrf_install_timeout(struct device *dev,
+				   struct wdt_timeout_cfg *cfg)
+{
+	nrfx_err_t err_code;
+	ARG_UNUSED(dev);
+
+	err_code = nrfx_wdt_channel_alloc(&m_ch[m_alloc_ch_num]);
+
+	if (err_code == NRFX_ERROR_NO_MEM) {
+		return -ENOMEM;
+	}
+
+	cfg->handle = &m_ch[m_alloc_ch_num];
```

Comment:
Now I see that we misunderstood the API a bit. This field was not originally supposed to be filled by the driver (see [here](https://github.com/zephyrproject-rtos/zephyr/pull/1260/files#diff-eb04a3826c198cc49f02f288abf3d859R140)). But I think such approach is troublesome because driver would need to do some mapping of the `handle` pointer to a proper watchdog channel and this will be ineffective. It will be better to use what you proposed earlier - to return from this function the index of the activated channel, and then expect this index to be passed to `wdt_feed`.

Reply:
Well, that's point to discuss. Feed function needs information about channel to be feeded. Now we have two points where we can do it. Solution based on pointer to struct is more generic imo. Now it is not completely compatible with documentation.
Surely it can be done using integer value returned by `install_timeout` function (second solution). However it requires a little changes in implementation.
I will align doc to choosen solution.

Reply:
Okay, My proposal is to return index of allocated watchdog channel in `install_timeout` function and use it to deal with `feed` function. Index can be easily translated inside the shim layer.
There is only one point to discuss. It requires one change in doc. I mean values started from zero are valid and point to channel index instead of pointing directly to "success" return code:
https://github.com/zephyrproject-rtos/zephyr/pull/1260/files#diff-eb04a3826c198cc49f02f288abf3d859R230

Reply:
Returning the index in `install_timeout` is fine.